File: launcher.cpp
#include <openssl/x509.h>
#include <openssl/x509_vfy.h>
#include <openssl/pem.h>
#include <openssl/rsa.h>
#include <openssl/evp.h>
#include <openssl/bn.h>
#include <openssl/err.h>
#include <openssl/x509v3.h>
#include <openssl/rand.h>
#include <openssl/ocsp.h>
#include <iostream>
#include <string>
#include <ctime>
#include <map>

#define CERTS_DIR "certs"
#define ROOT_KEY_FILE CERTS_DIR "/root_key.pem"
#define ROOT_CERT_FILE CERTS_DIR "/root_cert.pem"
#define CRL_FILE CERTS_DIR "/crl.pem"
#define CERT_DIR CERTS_DIR "/certs"

// Error handling macro
#define handle_error(msg) do { perror(msg); exit(1); } while (0)

using namespace std;

// Get current time
time_t get_current_time() {
    time_t now = time(0);
    return now;
}

// Self-signed root certificate creation
void generate_root_certificate(int days_valid = 3650) {
    RSA *rsa = RSA_new();
    BIGNUM *bn = BN_new();
    BN_set_word(bn, RSA_F4);
    if (RSA_generate_key_ex(rsa, 2048, bn, NULL) != 1) handle_error("RSA key generation failed");

    EVP_PKEY *pkey = EVP_PKEY_new();
    EVP_PKEY_assign_RSA(pkey, rsa);

    X509 *x509 = X509_new();
    X509_set_version(x509, 2);  // X509 version 3
    ASN1_INTEGER_set(X509_get_serialNumber(x509), 1);  // Serial number is set to 1 for root
    X509_gmtime_adj(X509_get_notBefore(x509), 0);  // Valid from now
    X509_gmtime_adj(X509_get_notAfter(x509), days_valid * 24 * 60 * 60);  // Expiration date

    X509_set_pubkey(x509, pkey);

    X509_NAME *name = X509_get_subject_name(x509);
    X509_NAME_add_entry_by_txt(name, "C", MBSTRING_ASC, (unsigned char *)"RO", -1, -1, 0);
    X509_NAME_add_entry_by_txt(name, "O", MBSTRING_ASC, (unsigned char *)"MyCA", -1, -1, 0);
    X509_NAME_add_entry_by_txt(name, "CN", MBSTRING_ASC, (unsigned char *)"MyRootCA", -1, -1, 0);
    X509_set_issuer_name(x509, name);

    if (X509_sign(x509, pkey, EVP_sha256()) == 0) handle_error("Error signing root certificate");

    FILE *key_file = fopen(ROOT_KEY_FILE, "wb");
    if (!key_file) handle_error("Unable to open root key file for writing");
    PEM_write_PrivateKey(key_file, pkey, NULL, NULL, 0, NULL, NULL);
    fclose(key_file);

    FILE *cert_file = fopen(ROOT_CERT_FILE, "wb");
    if (!cert_file) handle_error("Unable to open root cert file for writing");
    PEM_write_X509(cert_file, x509);
    fclose(cert_file);

    X509_free(x509);
    EVP_PKEY_free(pkey);
    BN_free(bn);

    cout << "Root certificate generated successfully." << endl;
}

// Generate a certificate signed by the root certificate
void generate_certificate(const string &domain, const string &issuer_cert_file, const string &issuer_key_file, int days_valid = 365) {
    // Load issuer private key
    FILE *key_file = fopen(issuer_key_file.c_str(), "rb");
    if (!key_file) handle_error("Failed to load issuer private key");
    EVP_PKEY *issuer_pkey = PEM_read_PrivateKey(key_file, NULL, NULL, NULL);
    fclose(key_file);
    if (!issuer_pkey) handle_error("Failed to load issuer private key");

    // Load issuer certificate
    FILE *cert_file = fopen(issuer_cert_file.c_str(), "rb");
    if (!cert_file) handle_error("Failed to load issuer certificate");
    X509 *issuer_cert = PEM_read_X509(cert_file, NULL, NULL, NULL);
    fclose(cert_file);
    if (!issuer_cert) handle_error("Failed to load issuer certificate");

    // Generate RSA key for the new certificate
    RSA *rsa = RSA_new();
    BIGNUM *bn = BN_new();
    BN_set_word(bn, RSA_F4);
    if (RSA_generate_key_ex(rsa, 2048, bn, NULL) != 1) handle_error("RSA key generation failed");

    EVP_PKEY *pkey = EVP_PKEY_new();
    EVP_PKEY_assign_RSA(pkey, rsa);

    // Create the X509 certificate
    X509 *x509 = X509_new();
    X509_set_version(x509, 2);
    ASN1_INTEGER_set(X509_get_serialNumber(x509), 1);  // Serial number for this certificate
    X509_gmtime_adj(X509_get_notBefore(x509), 0);  // Valid from now
    X509_gmtime_adj(X509_get_notAfter(x509), days_valid * 24 * 60 * 60);  // Expiration date

    X509_set_pubkey(x509, pkey);

    // Set the subject name (CN should be the domain)
    X509_NAME *name = X509_get_subject_name(x509);
    X509_NAME_add_entry_by_txt(name, "C", MBSTRING_ASC, (unsigned char *)"RO", -1, -1, 0);
    X509_NAME_add_entry_by_txt(name, "O", MBSTRING_ASC, (unsigned char *)"MyCA", -1, -1, 0);
    X509_NAME_add_entry_by_txt(name, "CN", MBSTRING_ASC, (unsigned char *)domain.c_str(), -1, -1, 0);
    X509_set_issuer_name(x509, X509_get_subject_name(issuer_cert));

    // Sign the certificate with the issuer's private key
    if (X509_sign(x509, issuer_pkey, EVP_sha256()) == 0) handle_error("Error signing certificate");

    // Save the signed certificate to file
    string cert_file_name = CERT_DIR "/" + domain + "_cert.pem";
    FILE *cert_out = fopen(cert_file_name.c_str(), "wb");
    if (!cert_out) handle_error("Unable to open certificate file for writing");
    PEM_write_X509(cert_out, x509);
    fclose(cert_out);

    // Save the private key to file
    string key_file_name = CERT_DIR "/" + domain + "_key.pem";
    FILE *key_out = fopen(key_file_name.c_str(), "wb");
    if (!key_out) handle_error("Unable to open key file for writing");
    PEM_write_PrivateKey(key_out, pkey, NULL, NULL, 0, NULL, NULL);
    fclose(key_out);

    // Clean up
    X509_free(x509);
    EVP_PKEY_free(pkey);
    EVP_PKEY_free(issuer_pkey);
    X509_free(issuer_cert);
    BN_free(bn);

    cout << "Certificate for " << domain << " generated and signed by root." << endl;
}

// Function to create a CRL (Certificate Revocation List) and add a revoked certificate
void create_crl(const string &revoked_cert_file_path) {
    // Create a new CRL object
    X509_CRL *crl = X509_CRL_new();
    if (!crl) handle_error("Failed to create CRL");

    // Set the CRL version
    X509_CRL_set_version(crl, 1);

    // Set the CRL issuer (should be the root certificate subject)
    FILE *root_cert_file = fopen(ROOT_CERT_FILE, "rb");
    X509 *root_cert = PEM_read_X509(root_cert_file, NULL, NULL, NULL);
    fclose(root_cert_file);
    X509_NAME *issuer_name = X509_get_subject_name(root_cert);
    X509_CRL_set_issuer_name(crl, issuer_name);

    // Set the last update time
    ASN1_TIME *last_update = ASN1_TIME_new();
    ASN1_TIME_set(last_update, get_current_time());
    X509_CRL_set_lastUpdate(crl, last_update);

    // Set the next update time
    ASN1_TIME *next_update = ASN1_TIME_new();
    ASN1_TIME_set(next_update, get_current_time() + 86400 * 365);  // 1 year from now
    X509_CRL_set_nextUpdate(crl, next_update);

    // Add the revoked certificate to the CRL
    FILE *revoked_cert_file = fopen(revoked_cert_file_path.c_str(), "rb"); // Renamed the local variable here
    X509 *revoked_cert = PEM_read_X509(revoked_cert_file, NULL, NULL, NULL);
    fclose(revoked_cert_file);
    X509_REVOKED *revoked_entry = X509_REVOKED_new();
    ASN1_TIME *revoked_time = ASN1_TIME_new();
    ASN1_TIME_set(revoked_time, get_current_time());  // Revocation time

    X509_REVOKED_set_serialNumber(revoked_entry, X509_get_serialNumber(revoked_cert));
    X509_REVOKED_set_revocationDate(revoked_entry, revoked_time);
    X509_CRL_add0_revoked(crl, revoked_entry);

    // Save the CRL to file
    FILE *crl_out = fopen(CRL_FILE, "wb");
    if (!crl_out) handle_error("Unable to open CRL file for writing");
    PEM_write_X509_CRL(crl_out, crl);
    fclose(crl_out);

    ASN1_TIME_free(last_update);
    ASN1_TIME_free(next_update);
    ASN1_TIME_free(revoked_time);
    X509_CRL_free(crl);
    X509_free(revoked_cert);

    cout << "CRL created and certificate revoked." << endl;
}


// Function to check certificate expiration and renew if necessary
void renew_certificate_if_needed(const string &domain, const string &issuer_cert_file, const string &issuer_key_file) {
    // Load the existing certificate to check its expiration
    string cert_file_name = CERT_DIR "/" + domain + "_cert.pem";
    FILE *cert_file = fopen(cert_file_name.c_str(), "rb");
    X509 *existing_cert = PEM_read_X509(cert_file, NULL, NULL, NULL);
    fclose(cert_file);
    if (!existing_cert) handle_error("Failed to load existing certificate");

    // Check expiration date
    ASN1_TIME *not_after = X509_get_notAfter(existing_cert);
    time_t current_time = get_current_time();
    if (ASN1_TIME_diff(NULL, NULL, not_after, NULL) < 0) {
        // If expired, renew the certificate
        generate_certificate(domain, issuer_cert_file, issuer_key_file);
    }

    X509_free(existing_cert);
}

int main() {

    system("mkdir -p certs/certs");
    generate_root_certificate();

    // Generate a certificate for a domain (example: "example.com")
    generate_certificate("example.com", ROOT_CERT_FILE, ROOT_KEY_FILE);

    // Create CRL and revoke a certificate (for testing purposes)
    create_crl(CERT_DIR "/example.com_cert.pem");

    // Check if the certificate is expired and renew if necessary
    renew_certificate_if_needed("example.com", ROOT_CERT_FILE, ROOT_KEY_FILE);

    return 0;
}

File: launcher.py
from cryptography import x509
from cryptography.x509.oid import NameOID
from cryptography.hazmat.primitives.asymmetric import rsa
from cryptography.hazmat.primitives import serialization, hashes
from datetime import datetime, timedelta
import os

# Directory to store certificates
CERTS_DIR = "certs"
os.makedirs(CERTS_DIR, exist_ok=True)

CRL_FILE = f"{CERTS_DIR}/crl.pem"

# Generate a self-signed root certificate
def generate_root_certificate():
    key = rsa.generate_private_key(public_exponent=65537, key_size=2048)
    subject = issuer = x509.Name([
        x509.NameAttribute(NameOID.COUNTRY_NAME, "RO"),
        x509.NameAttribute(NameOID.STATE_OR_PROVINCE_NAME, "Bucharest"),
        x509.NameAttribute(NameOID.ORGANIZATION_NAME, "MyCA"),
        x509.NameAttribute(NameOID.COMMON_NAME, "MyRootCA")
    ])
    certificate = (
        x509.CertificateBuilder()
        .subject_name(subject)
        .issuer_name(issuer)
        .public_key(key.public_key())
        .serial_number(x509.random_serial_number())
        .not_valid_before(datetime.utcnow())
        .not_valid_after(datetime.utcnow() + timedelta(days=3650))
        .add_extension(x509.BasicConstraints(ca=True, path_length=None), critical=True)
        .sign(private_key=key, algorithm=hashes.SHA256())
    )
    # Save root key & certificate
    with open(f"{CERTS_DIR}/root_key.pem", "wb") as f:
        f.write(key.private_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PrivateFormat.TraditionalOpenSSL,
            encryption_algorithm=serialization.NoEncryption()
        ))
    with open(f"{CERTS_DIR}/root_cert.pem", "wb") as f:
        f.write(certificate.public_bytes(serialization.Encoding.PEM))
    print("Root certificate generated successfully.")

# Generate a new certificate signed by the root
def issue_certificate(common_name):
    with open(f"{CERTS_DIR}/root_key.pem", "rb") as f:
        root_key = serialization.load_pem_private_key(f.read(), password=None)
    with open(f"{CERTS_DIR}/root_cert.pem", "rb") as f:
        root_cert = x509.load_pem_x509_certificate(f.read())
    key = rsa.generate_private_key(public_exponent=65537, key_size=2048)
    subject = x509.Name([
        x509.NameAttribute(NameOID.COMMON_NAME, common_name),
    ])
    certificate = (
        x509.CertificateBuilder()
        .subject_name(subject)
        .issuer_name(root_cert.subject)
        .public_key(key.public_key())
        .serial_number(x509.random_serial_number())
        .not_valid_before(datetime.utcnow())
        .not_valid_after(datetime.utcnow() + timedelta(days=365))
        .add_extension(x509.SubjectKeyIdentifier.from_public_key(key.public_key()), critical=False)
        .add_extension(x509.AuthorityKeyIdentifier.from_issuer_public_key(root_cert.public_key()), critical=False)
        .sign(private_key=root_key, algorithm=hashes.SHA256())
    )
    with open(f"{CERTS_DIR}/{common_name}_key.pem", "wb") as f:
        f.write(key.private_bytes(
            encoding=serialization.Encoding.PEM,
            format=serialization.PrivateFormat.TraditionalOpenSSL,
            encryption_algorithm=serialization.NoEncryption()
        ))
    with open(f"{CERTS_DIR}/{common_name}_cert.pem", "wb") as f:
        f.write(certificate.public_bytes(serialization.Encoding.PEM))
    print(f"Certificate issued for {common_name} with serial number {certificate.serial_number} and expiration {certificate.not_valid_after}.")

# Check if a certificate is expired and renew it
def renew_certificate(common_name):
    cert_path = f"{CERTS_DIR}/{common_name}_cert.pem"
    if not os.path.exists(cert_path):
        print(f"Certificate for {common_name} does not exist.")
        return
    with open(cert_path, "rb") as f:
        cert = x509.load_pem_x509_certificate(f.read())
    if cert.not_valid_after < datetime.utcnow():
        print(f"Certificate for {common_name} is expired. Renewing...")
        issue_certificate(common_name)
    else:
        print(f"Certificate for {common_name} is still valid until {cert.not_valid_after}.")

# Revoke a certificate
def revoke_certificate(serial_number):
    revoked_cert = x509.RevokedCertificateBuilder()
    revoked_cert = revoked_cert.serial_number(serial_number)
    revoked_cert = revoked_cert.revocation_date(datetime.utcnow())
    revoked_cert = revoked_cert.build()
    crl_builder = x509.CertificateRevocationListBuilder()
    crl_builder = crl_builder.issuer_name(x509.Name([
        x509.NameAttribute(NameOID.COMMON_NAME, "MyRootCA")
    ]))
    crl_builder = crl_builder.last_update(datetime.utcnow())
    crl_builder = crl_builder.next_update(datetime.utcnow() + timedelta(days=30))
    crl_builder = crl_builder.add_revoked_certificate(revoked_cert)
    with open(f"{CERTS_DIR}/root_key.pem", "rb") as f:
        root_key = serialization.load_pem_private_key(f.read(), password=None)
    crl = crl_builder.sign(private_key=root_key, algorithm=hashes.SHA256())
    with open(CRL_FILE, "wb") as f:
        f.write(crl.public_bytes(serialization.Encoding.PEM))
    print(f"Certificate with serial number {serial_number} revoked.")

if __name__ == "__main__":
    generate_root_certificate()
    issue_certificate("example.com")
    renew_certificate("example.com")
    revoke_certificate(123456789)  # Example serial number for revocation
